DEVELOPERS.md

```diff
@@ -62,6 +62,159 @@ running jobs from private GitHub repos, and/or you want to exercise the full tes
 suite (some tests in `tests/lib/test_git.py` are skipped if this environment variable
 is missing).
 
+## Operating principles
+
+The project retrieves jobs to be run from an [OpenSAFELY job
+server](https://github.com/opensafely-core/job-server) by polling the job server.
+
+Jobs belong to a `workspace`. This describes the git repo containing the
+OpenSAFELY-compliant project under execution; the git branch, and kind of
+database to use. The workspace also acts as a kind of namespace for
+partitioning outputs of its jobs.
+
+An OpenSAFELY-compliant repo must provide a `project.yaml` file which
+describes how a requested job should be converted into a command (& arguments)
+that can be run in a subprocess on the secure server.  It incorporates the idea
+of dependencies, so an action that generates a chart might depend on an action
+that extracts data from the database *for* that chart.  See the
+[Actions reference](https://docs.opensafely.org/actions-intro/) for more information.
+
+An action can define `outputs`; these are persisted on disk and made available
+to subsequent actions in the workspace, and users who have permission to log
+into the server and view the raw files.
+
+The runner takes care of executing dependencies in order. By default, it skips
+re-running a dependency whose previous run produced output that still exists in
+the production environment.  The runner also reports status back to the job
+server, redacting possibly-sensitive information.
+
+The runner is bundled as part of the [opensafely-cli][cli] tool so users
+can test their actions locally.
+
+[cli]: https://github.com/opensafely-core/opensafely-cli
+
+### Job structure
+
+The job server serves jobs as JSON in the following format. First, a job must
+belong to a workspace:
+
+```json
+{
+    "workspace": {
+        "name": "my workspace",
+        "repo": "https://github.com/opensafely/job-integration-tests",
+        "branch": "main",
+        "db": "full"
+    }
+}
+```
+Possible values for `"db"` are "full", "slice", and "dummy".
```

Comment:
This is definitely not accurate any more; from config, valid database names are "default" and "include_t1oo" (but "include_t1oo"  only applies to TPP, atm at least)

Reply:
fixed, maybe this is also called `database_name` now, I'm just checking

Reply:
renamed, and some wider follow-up changes too